### PR TITLE
Fix EZP-24898: Role versioning breaks data BC in legacy storage engine

### PR DESCRIPTION
**Pull request by lolautruche**
> https://jira.ez.no/browse/EZP-24898

#1376 introduced role versioning, but not by following legacy behavior.

Consequences are:
* Data BC is broken between Platform and Legacy, which is annoying for migration.
* In `ezpolicy` table, `role_id` column holds a reference to the parent role, being a draft or not. A role draft is a normal role *decorating* another one.
* Role draft has policies with no reference to original policies, which is annoying especially for policy editing in PlatformUI.

This PR restores the behavior from legacy kernel.

## TODO
* [x] Change legacy storage engine gateway to make it behave like legacy (for roles and policies)
* [x] Change SPI accordingly
* [x] Add `RoleService::loadRoleDraftByRoleId()` since there is no way to load a draft from a published role any more
* [x] Remove update SQL scripts
* [x] Introduce a `PolicyDraft` value object which acts like `RoleDraft`
* [x] Add tests for the new API method
* [x] Add tests for new SPI method

## Open questions
* We had an update SQL script for both MySQL and PostGres, but we actually need to revert it. For now I just removed the scripts. Is this OK? We could document the revert in the release notes.

### CODE REVIEW

**Comment by lolautruche** [comment]
Review ping @andrerom @bdunogier @pspanja @glye @yannickroger

**Comment by glye** [comment]
`PolicyDraft` class is missing..?

**Comment by lolautruche** [comment]
@glye No I chose not to create a `PolicyDraft` class as it complicated things a lot (I tried). 
Rant from @ezrobot was due to a leftover.

File: eZ/Publish/Core/Persistence/Legacy/Tests/User/UserHandlerTest.php [hunk 1]
```diff
@@ -325,7 +325,7 @@ public function testLoadRoleWithPolicies()
                         'module' => 'foo',
                         'function' => 'bar',
                         'limitations' => '*',
-                        'status' => APIRole::STATUS_DEFINED,
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/Tests/User/UserHandlerTest.php]: why is this removed? when not removed form api?

File: eZ/Publish/API/Repository/Values/User/Policy.php [hunk 2]
```diff
@@ -56,6 +57,14 @@
     protected $function;
 
     /**
+     * Original policy ID the policy was created from.
+     * Mostly used when role status is Role::STATUS_DRAFT.
+     *
+     * @var mixed
+     */
+    protected $originalId;
```
Comment (andrerom) [inline eZ/Publish/API/Repository/Values/User/Policy.php]: This does not really belong here, it should ideally be a PolicyDraft, and return that from API whenever this id is set from db (so most likely need to use -1 for new draft).

File: eZ/Publish/SPI/Persistence/User/Policy.php [hunk 3]
```diff
@@ -31,13 +31,14 @@ class Policy extends ValueObject
...
 
     /**
-     * Status of the policy (legacy: "original_id").
+     * Only used when the role's status, current policy belongs to, is Role::STATUS_DRAFT.
+     * Original policy ID the draft was created from.
      *
      * @since 6.0
      *
-     * @var int One of Role::STATUS_DEFINED|Role::STATUS_DRAFT
+     * @var int
```
Comment (andrerom) [inline eZ/Publish/SPI/Persistence/User/Policy.php]: This is also null when not a draft right? And should document what value will be on newly created draft.

**Comment by lolautruche** [comment]
@andrerom @glye @bdunogier @pspanja Review needed, even if not completely finished :smile:

**Comment by lolautruche** [comment]
Also ping @emodric Since you were concerned about this BC break :wink:

**Comment by emodric** [comment]
@lolautruche I'll take a look on Monday, can't manage over the weekend, due to conference being held here in Zagreb.

Thanks for the effort :)

**Comment by andrerom** [comment]
@lolautruche is task list up to date? *review question*

**Comment by lolautruche** [comment]
Should be, yes

File: data/updates/6.0/mysql.sql [hunk 4]
```diff
@@ -1,10 +0,0 @@
---
--- The `ezrole` table has changed. `version` has been added to the primary key, after the `id`. This is because role
--- drafts are now handled differently. A draft now has the same `id` as the published role, and `version` decides which
--- is which. You should make sure you don't have any role drafts in the database when upgrading. To find drafts, select
--- entries in the `ezrole` table where `version` is not zero. Before this change, the `version` of a draft equals the
--- `id` of the role. The safe way to get rid of stray role drafts is to edit the role in question, and then cancel
--- editing.
---
-
-ALTER TABLE ezrole DROP PRIMARY KEY, ADD PRIMARY KEY(id,version);
```
Comment (andrerom) [inline data/updates/6.0/mysql.sql]: We will need entry about this in 15.09 upgrade doc, for users that was using 15.07 *(or master)*.

Comment (lolautruche) [inline data/updates/6.0/mysql.sql]: Yes, hence the open Question on the PR :-)

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 5]
```diff
@@ -47,15 +47,21 @@ public function __construct(DatabaseHandler $handler)
...
     public function createRole(Role $role)
     {
-        // Role ID is set when creating a draft from an existing role
+        // Role original ID is set when creating a draft from an existing role
+        if ($role->status === Role::STATUS_DRAFT && $role->id) {
+            $roleOriginalId = $role->id;
+        } elseif ($role->status === Role::STATUS_DRAFT) {
+            $roleOriginalId = -1;
+        } else {
+            $roleOriginalId = Role::STATUS_DEFINED;
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: For better readability this should either use `0`, or a constant that makes it clear, like `DEFINED_ROLE_EMPTY_ORIGINAL_ID`. Same point goes for use of `-1` above.

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 6]
```diff
@@ -89,6 +97,18 @@ public function createRole(Role $role)
     public function loadRole($roleId, $status = Role::STATUS_DEFINED)
     {
         $query = $this->handler->createSelectQuery();
+        if ($status === Role::STATUS_DEFINED) {
+            $draftCondition = $query->expr->eq(
+                $this->handler->quoteColumn('version', 'ezrole'),
+                $query->bindValue($status, null, \PDO::PARAM_INT)
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: same as above, `0` or const for readability.

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 7]
```diff
@@ -89,6 +97,18 @@ public function createRole(Role $role)
...
+            );
+        } else {
+            $draftCondition = $query->expr->neq(
+                $this->handler->quoteColumn('version', 'ezrole'),
+                $query->bindValue(Role::STATUS_DEFINED, null, \PDO::PARAM_INT)
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: same

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 8]
```diff
@@ -155,6 +166,18 @@ public function loadRole($roleId, $status = Role::STATUS_DEFINED)
     public function loadRoleByIdentifier($identifier, $status = Role::STATUS_DEFINED)
     {
         $query = $this->handler->createSelectQuery();
+        if ($status === Role::STATUS_DEFINED) {
+            $roleVersionCondition = $query->expr->eq(
+                $this->handler->quoteColumn('version', 'ezrole'),
+                $query->bindValue($status, null, \PDO::PARAM_INT)
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: same :)

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 9]
```diff
@@ -155,6 +166,18 @@ public function loadRole($roleId, $status = Role::STATUS_DEFINED)
...
+            );
+        } else {
+            $roleVersionCondition = $query->expr->neq(
+                $this->handler->quoteColumn('version', 'ezrole'),
+                $query->bindValue($status, null, \PDO::PARAM_INT)
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: same :)

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 10]
```diff
@@ -220,6 +289,18 @@ public function loadRoleByIdentifier($identifier, $status = Role::STATUS_DEFINED
     public function loadRoles($status = Role::STATUS_DEFINED)
     {
         $query = $this->handler->createSelectQuery();
+        if ($status === Role::STATUS_DEFINED) {
+            $roleVersionCondition = $query->expr->eq(
+                $this->handler->quoteColumn('version', 'ezrole'),
+                $query->bindValue($status, null, \PDO::PARAM_INT)
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: same

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php [hunk 11]
```diff
@@ -220,6 +289,18 @@ public function loadRoleByIdentifier($identifier, $status = Role::STATUS_DEFINED
...
+            );
+        } else {
+            $roleVersionCondition = $query->expr->neq(
+                $this->handler->quoteColumn('version', 'ezrole'),
+                $query->bindValue($status, null, \PDO::PARAM_INT)
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: same

**Comment by andrerom** [comment]
Looks good, and as for the "status" review comments, I see now that it is rather consistent as is, so maybe we should agree to let that be. what's your POV?

**Comment by lolautruche** [comment]
For the status, the Role constant is used wherever possible. Only thing was the - 1 for newly created roles. I didn't add a constant as I considered this legacy storage specific.

File: eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php (hunk 5 above)
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/User/Role/Gateway/DoctrineDatabase.php]: Alternative, just add inline comment there making it clear Role::STATUS_DEFINED equals 0, so it is clear from reading these lines how db values will look like.

**Comment by andrerom** [comment]
Ok, removed all those status comments besides the first one, as it has a mix of different values set on roleOriginalId it would be nice to easily see what value will be in db by just reading it.

File: data/updates/6.0/mysql.sql (hunk 4 above)
Comment (andrerom) [inline data/updates/6.0/mysql.sql]: So the answer is; move it to `6.0/unstable/dbupdate-5.4.0-to-6.0.0alpha7.sql` and create one for reverting it for 15.07 users in `6.0/unstable/dbupdate-6.0.0alpha7-to-6.0.0beta1.sql` ;)

**Comment by glye** [comment]
There are still some `PolicyDraft` references in the PR, assuming it will implemented or fully removed.
OK to remove the update scripts if revert is documented. +1 when Travis is at peace.

**Comment by lolautruche** [comment]
OK, should be fine now, all remarks taken into account

**Comment by andrerom** [comment]
+1

File: eZ/Publish/API/Repository/Values/User/Policy.php [hunk 12]
```diff
@@ -19,6 +19,7 @@
  * @property-read mixed $roleId the role id this policy belongs to
  * @property-read string $module Name of module, associated with the Policy
  * @property-read string $function  Name of the module function Or all functions with '*'
+ * @property-read mixed $originalId Original policy ID the policy was created from.
```
Comment (pspanja) [inline eZ/Publish/API/Repository/Values/User/Policy.php]: AFAICS this is not implemented in the concrete class.

**Comment by pspanja** [comment]
> No I chose not to create a PolicyDraft class as it complicated things a lot (I tried). 

Was adding `status` column to `ezpolicy` table an option here or?

**Comment by lolautruche** [comment]
Might be, but again we'll break data BC.
Finally with `PolicyDraft` we're good.

File: eZ/Publish/API/Repository/Values/User/Policy.php (hunk 12 above)
Comment (lolautruche) [inline eZ/Publish/API/Repository/Values/User/Policy.php]: Thanks. Fixed now.

File: doc/specifications/rest/REST-API-V2.rst [hunk 13]
```diff
@@ -5370,6 +5371,33 @@ Load Role
...
+.. code:: http
+
+          HTTP/1.1 200 OK
+          Accept-Patch:  application/vnd.ez.api.RoleInput+(json|xml)
+          ETag: "<Etag>"
+          Content-Type: <depending on accept header>
+          Content-Length: <length>
+
+.. parsed-literal::
+Role_
```
Comment (bdunogier) [inline doc/specifications/rest/REST-API-V2.rst]: Does it return a Role, or a RoleDraft object ?

File: eZ/Publish/Core/REST/Server/Controller/Role.php [hunk 14]
```diff
@@ -204,30 +212,35 @@ public function updateRoleDraft($roleId, Request $request)
...
+            // We might want a newly created role, so try to load it by its ID.
+            // loadRoleDraft() might throw a NotFoundException (wrong $roleId). If so, let it bubble up.
+            $roleDraft = $this->roleService->loadRoleDraft($roleId);
+        }
 
-        $this->roleService->publishRoleDraft(
-            $this->roleService->loadRoleDraft($roleId),
-            $this->mapToUpdateStruct($createStruct)
-        );
+        $this->roleService->publishRoleDraft($roleDraft);
```
Comment (bdunogier) [inline eZ/Publish/Core/REST/Server/Controller/Role.php]: The action should return something :-)

It would be something like return a `PublishedRole` value object or something, that has the Role as a property. Then the visitor would send a 200, with a Location header to the published Role (so that API users can follow the link).

File: eZ/Publish/Core/Repository/Values/User/RoleDraft.php [hunk 15]
```diff
@@ -64,7 +64,7 @@ public function __isset($property)
     /**
      * Returns the list of policies of this role.
      *
-     * @return \eZ\Publish\API\Repository\Values\User\Policy[]
+     * @return \eZ\Publish\Core\Repository\Values\User\PolicyDraft[]
```
Comment (bdunogier) [inline eZ/Publish/Core/Repository/Values/User/RoleDraft.php]: API or Core ?

**Comment by bdunogier** [comment]
Good enough. The REST spec might need some polishing, but it's good enough.

+1

File: eZ/Publish/Core/Repository/Values/User/RoleDraft.php (hunk 15 above)
Comment (andrerom) [inline eZ/Publish/Core/Repository/Values/User/RoleDraft.php]: API, or inherit doc.
